### ISSUE TITLE
New Plugin Manager

### ISSUE THREAD

**Issue by AndrewBelt** [VCVRack/Rack#789]
Upon clicking "Update plugins":
- Will use the existing method of checking the slugs and versions of all loaded plugins and comparing them with the external API.
- Downloads needed ZIP files to `plugins/`. Does *not* unzip them.

At startup:
- If a ZIP file exists in `plugins/`, unzip it (overwriting files) and delete the ZIP file.
- A bundled `Fundamental-*.zip` package will be unzipped in to the *local* plugin directory, *if* no `Fundamental` folder exists (this is a special case of the above).
- Load all plugins
- Display error if a folder in `plugins/` exists which either doesn't have a `plugin.{dll,dylib,so}` file, or it fails to load.

The global `plugins/` directory will be removed in Rack 0.6.

**Comment by cschol** [comment]
Sounds great! A few thoughts:
- If a ZIP file exists, the old plugin folder should be deleted before the new ZIP is unzipped, i.e. a new version should *completely replace* the older version.
- If ZIP extraction fails? Instead of deleting the older folder, you could *rename* the old plugin folder temporarily to "plugin_old", then extract the new one and if the that succeeds, remove the "plugin_old" folder. If extraction fails, restore the old folder.
- If a ZIP file's SHA check fails, the ZIP file should be deleted.
- You could keep the latest version of the ZIP files around (in a subfolder in `plugins/`) and use them to determine if a plugin needs update by verifying the SHA of the ZIP file against the SHA from the manifest. This would cover the case that a plugin author updates the plugin, but does not rev the version. The SHA must be updated anyway by the author since otherwise the SHA check fails, but someone might think not updating the version is OK.

**Comment by AndrewBelt** [comment]
1. I don't really want to do this because it might bite people. I believe there's nothing wrong with overwriting, although it might leave a mess of useless files. If the folder is deleted, it might remove developers' source code, files saved in the plugin directory (for whatever reason), etc.
2. If a folder fails to unzip, I'll just display an error in a similar way as if a plugin fails to load.
3. That's how it'll work. Added to list.
4. If we do that, let's go all the way and remove versions from the plugin entirely. Thoughts on what it would take to do that? Leaving a copy of `PluginSlug-VERSION-SLUG.zip` around just for bookkeeping? Placing a `plugin.json` file in the plugin folder after extracting? I don't like either of those ideas.

**Comment by cschol** [comment]
1. I see your point and I think you are right. I didn't think of plugins saving settings or other resources inside of their directory that might want to be migrated. Developer's source code inside of the plugins directory has other implications: at worst you loose unstaged changes by overwriting them with an extracted file by accidentally hitting Update while developing. 
It would be nice if you could keep your dev plugin folder separate from the managed plugin folder. E.g. an environment variable `RACK_PLUGIN_DIR` could be appended to the plugin search path and Rack could look for plugins there if the variable is defined. Analogous to `PYTHON_PATH`. One would have to define `RACK_DIR` also. Or maybe just clone your source into `slug.git` instead of using just the slug name.
2. OK. Now, the plugin load might really fail to load if the `plugin.dll` wasn't extracted or it could have other issues, like missing updated `res` files. Failed extraction is probably the worst of errors that can happen. It potentially leaves the plugin (and Rack?) in a bad state. Would the error advise the user to delete the plugin folder and try again?
3. *(just preserving the numbering)*
4. One way is to calculate a cumulative checksum of all files in the extracted plugin. That would require the plugin developer to provide the same in the plugin manifest. If that functionality (script?) was provided as part of the Rack plugin framework and printed out when `make dist` is called, this even would be feasible for developers.
I personally don't see any real downsides (besides disk space) of keeping the *latest* ZIP file around (not all versions). There is another case where that could be useful: it gives you a state to roll back to when things go awry within the plugin, *OFFLINE* without requiring network access: if the ZIP is there and the slug folder is not, unzip. So if I delete my plugin folder, it would be recreated from the ZIP next startup.

**Comment by almostEric** [comment]
It looks like the plugin manager just checks to see if version of plugin on server is different than installed, rather than newer. I'm testing my own plugins, and the update plugin button enables - clicking it overwrites my plugin with the older version. Not sure if that is by design (and admittedly not a common use case)

**Comment by AndrewBelt** [comment]
@almostEric Correct, this allows for versions like `0.6.1a`, `0.6.42rc1`. This shouldn't be a problem if you 1) build from source without `make RELEASE=1`, or 2) you uncheck your own plugin from the plugin manager.

**Comment by cschol** [comment]
Looks like the SHA check was not implemented for the new plugin manager?
